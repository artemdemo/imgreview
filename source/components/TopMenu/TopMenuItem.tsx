import React from 'react';
import styled from 'styled-components';
import Icon from '../Icon/Icon';
import SubMenu, { TSubmenuData } from './SubMenu';
import ClearButton from '../Button/ClearButton';
import * as styleVars from '../../styles/variables';

const horizontalPadding = 4;

const MainMenuItemSty__Content = styled.span`
    flex-grow: 1;
    min-width: ${30 - 2 * horizontalPadding}px;
`;

const MainMenuItemSty__Caret = styled.span`
    flex-grow: 0;
    padding-left: 5px;
`;

const MainMenuItemSty__Submenu = styled.div`
    position: absolute;
    top: 100%;
    display: none;
    padding-top: 5px;
    z-index: ${styleVars.mainMenuZIndex};
`;

const MainMenuItemSty = styled(ClearButton)`
    background-color: ${styleVars.mainMenuColor};
    padding: ${horizontalPadding}px 6px;
    border: 1px solid ${styleVars.mainMenuItemBoderColor};
    border-radius: 3px;
    float: left;
    margin-right: 5px;
    outline: none;
    display: flex;
    position: relative;

    &:active, &:focus {
        outline: 0;
    }
    
    &:hover ${MainMenuItemSty__Submenu} {
        display: block;
    }
    
    ${props => props.disabled && `
        opacity: 0.4;
        cursor: not-allowed;
    `}
`;

type Props = {
    subMenu: TSubmenuData;
    disabled: boolean;
    onClick: (e?: any) => void;
};

class TopMenuItem extends React.PureComponent<Props> {
    static readonly defaultProps = {
        onClick: null,
        disabled: null,
        subMenu: [],
    };

    renderCaret() {
        const { subMenu } = this.props;
        if (subMenu.length > 0) {
            return (
                <MainMenuItemSty__Caret>
                    <Icon name='caret-down' />
                </MainMenuItemSty__Caret>
            );
        }

        return null;
    }

    renderSubMenu() {
        const { subMenu } = this.props;
        if (subMenu.length > 0) {
            return (
                <MainMenuItemSty__Submenu>
                    <SubMenu data={subMenu} />
                </MainMenuItemSty__Submenu>
            );
        }

        return null;
    }

    render() {
        const { disabled, onClick } = this.props;
        return (
            <MainMenuItemSty
                disabled={disabled}
                onClick={onClick}
                type='button'
            >
                <MainMenuItemSty__Content>
                    {this.props.children}
                </MainMenuItemSty__Content>
                {this.renderCaret()}
                {this.renderSubMenu()}
            </MainMenuItemSty>
        );
    }
}

export default TopMenuItem;
