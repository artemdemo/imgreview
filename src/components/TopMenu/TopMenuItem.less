@import '../../styles/variables.less';

.TopMenuItem__Content {
  flex-grow: 1;
  min-width: 22px;
  text-align: center;
  font-size: 95%;
}

.TopMenuItem__Caret {
  flex-grow: 0;
  padding-left: 5px;
}

.TopMenuItem__Submenu {
  position: absolute;
  display: none;
  top: 100%;
  padding-top: 5px;
  z-index: @main-menu-z-index;
}

.TopMenuItem__Submenu_open {
  display: block;
}
