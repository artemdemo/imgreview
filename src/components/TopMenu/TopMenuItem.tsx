import React from 'react';
import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faAngleDown } from '@fortawesome/free-solid-svg-icons';
import SubMenu, { TSubmenuData } from './SubMenu';
import MenuButton from './MenuButton';
import * as styleVars from '../../styles/variables';

const MainMenuItem__Content = styled.span`
    flex-grow: 1;
    min-width: 22px;
    text-align: center;
`;

const MainMenuItem__Caret = styled.span`
    flex-grow: 0;
    padding-left: 5px;
`;

const MainMenuItem__Submenu = styled.div`
    position: absolute;
    top: 100%;
    display: ${props => props.open ? 'block' : 'none'};
    padding-top: 5px;
    z-index: ${styleVars.mainMenuZIndex};
`;

type TProps = {
    subMenu: TSubmenuData;
    show: boolean;
    disabled: boolean;
    active?: boolean;
    open: boolean;
    href: string;
    title: string;
    onClick: (e?: any) => void;
    stopPropagation: boolean;
};

class TopMenuItem extends React.PureComponent<TProps> {
    static readonly defaultProps = {
        onClick: null,
        show: true,
        disabled: false,
        open: false,
        href: '',
        title: '',
        subMenu: [],
        stopPropagation: true,
    };

    hasSubmenu() {
        const { subMenu } = this.props;
        return subMenu.length > 0;
    }

    onClick = (e) => {
        const {onClick, stopPropagation} = this.props;
        if (stopPropagation) {
            // Parent <Menu> will blur shapes, but it will happened _after_ I add new arrow.
            // I don't want ot menu to handle blurring, since I want that new arrow will stay in focus.
            e.stopPropagation();
        }
        onClick && onClick(e);
    };

    renderCaret() {
        if (this.hasSubmenu()) {
            return (
                <MainMenuItem__Caret>
                    <FontAwesomeIcon icon={faAngleDown} />
                </MainMenuItem__Caret>
            );
        }

        return null;
    }

    renderSubMenu() {
        const { subMenu, open } = this.props;
        if (this.hasSubmenu()) {
            return (
                <MainMenuItem__Submenu open={open}>
                    <SubMenu data={subMenu} />
                </MainMenuItem__Submenu>
            );
        }

        return null;
    }

    render() {
        const { disabled, show, active, href, title } = this.props;
        if (show) {
            return (
                <MenuButton
                    disabled={disabled}
                    active={active}
                    onClick={this.onClick}
                    href={href}
                    title={title}
                    posRelative={this.hasSubmenu()}
                >
                    <MainMenuItem__Content>
                        {this.props.children}
                    </MainMenuItem__Content>
                    {this.renderCaret()}
                    {this.renderSubMenu()}
                </MenuButton>
            );
        }
        return null;
    }
}

export default TopMenuItem;
