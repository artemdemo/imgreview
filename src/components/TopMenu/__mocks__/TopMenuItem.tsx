import React from 'react';

const TopMenuItem = (props) => (
  <div onClick={props.onClick} data-mock="TopMenuItem">
    {props.children}
  </div>
);

export default TopMenuItem;
