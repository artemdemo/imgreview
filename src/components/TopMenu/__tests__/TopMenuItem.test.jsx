import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import TopMenuItem from '../TopMenuItem';

jest.mock('../SubMenu');
jest.mock('../MainItemWrap');

describe('TopMenuItem', () => {
    it('simple render', () => {
        const tree = renderer.create(
            <TopMenuItem />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should be disabled', () => {
        const tree = renderer.create(
            <TopMenuItem disabled />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should handle subMenu', () => {
        const wrapper = mount(
            <TopMenuItem
                subMenu={[
                    {text: 'some text - 1', value: 1},
                    {text: 'some text - 2', value: 2},
                ]}
            >
                Button text
            </TopMenuItem>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
