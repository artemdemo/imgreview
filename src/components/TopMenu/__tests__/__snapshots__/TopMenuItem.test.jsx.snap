// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopMenuItem should be disabled 1`] = `
<button
  className="sc-bdVaJa sc-ifAKCX cckwYG"
  disabled={true}
  onClick={null}
  type="button"
>
  <span
    className="sc-bwzfXH doRSpX"
  />
</button>
`;

exports[`TopMenuItem should handle subMenu 1`] = `
<TopMenuItem
  disabled={false}
  href=""
  onClick={null}
  open={false}
  subMenu={
    Array [
      Object {
        "text": "some text - 1",
        "value": 1,
      },
      Object {
        "text": "some text - 2",
        "value": 2,
      },
    ]
  }
>
  <Styled(styled.button)
    disabled={false}
    onClick={null}
    type="button"
  >
    <StyledComponent
      disabled={false}
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-ifAKCX",
            "isStatic": false,
            "lastClassName": "jzIrgm",
            "rules": Array [
              "
    ",
              "
    border: none;
    margin: 0;
    padding: 0;
    width: auto;
    overflow: visible;

    background: transparent;

    cursor: pointer;

    /* inherit font & color from ancestor */
    color: inherit;
    font: inherit;

    /* Normalize 'line-height'. Cannot be changed from 'normal' in Firefox 4+. */
    line-height: normal;

    /* Corrects font smoothing for webkit */
    -webkit-font-smoothing: inherit;
    -moz-osx-font-smoothing: inherit;

    /* Corrects inability to style clickable 'input' types in iOS */
    -webkit-appearance: none;
",
              "
",
              "
    ",
              "
    background-color: ",
              "#D3D3D3",
              ";
    padding: ",
              "4",
              "px 6px;
    border: 1px solid ",
              "#b8b8b8",
              ";
    border-radius: 3px;
    float: left;
    margin-right: 5px;
    outline: none;
    display: flex;
    position: relative;

    &:active, &:focus {
        outline: 0;
    }

    // @ts-ignore
    ",
              [Function],
              ";
",
              "
",
            ],
          },
          "displayName": "Styled(styled.button)",
          "foldedComponentIds": Array [
            "sc-bdVaJa",
          ],
          "render": [Function],
          "styledComponentId": "sc-ifAKCX",
          "target": "button",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      onClick={null}
      type="button"
    >
      <button
        className="sc-bdVaJa sc-ifAKCX jzIrgm"
        disabled={false}
        onClick={null}
        type="button"
      >
        <styled.span>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": true,
                  "lastClassName": "doRSpX",
                  "rules": Array [
                    "
    flex-grow: 1;
    min-width: ",
                    "22",
                    "px;
    text-align: center;
",
                  ],
                },
                "displayName": "styled.span",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <span
              className="sc-bwzfXH doRSpX"
            >
              Button text
            </span>
          </StyledComponent>
        </styled.span>
        <styled.span>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-htpNat",
                  "isStatic": true,
                  "lastClassName": "dzJpfP",
                  "rules": Array [
                    "
    flex-grow: 0;
    padding-left: 5px;
",
                  ],
                },
                "displayName": "styled.span",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-htpNat",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <span
              className="sc-htpNat dzJpfP"
            >
              <Icon
                name="caret-down"
              >
                <div
                  data-mock="Icon"
                />
              </Icon>
            </span>
          </StyledComponent>
        </styled.span>
        <styled.div
          open={false}
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bxivhb",
                  "isStatic": false,
                  "lastClassName": "gjrfiv",
                  "rules": Array [
                    "
    position: absolute;
    top: 100%;
    display: ",
                    [Function],
                    ";
    padding-top: 5px;
    z-index: ",
                    "10",
                    ";
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bxivhb",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            open={false}
          >
            <div
              className="sc-bxivhb gjrfiv"
              open={false}
            >
              <SubMenu
                data={
                  Array [
                    Object {
                      "text": "some text - 1",
                      "value": 1,
                    },
                    Object {
                      "text": "some text - 2",
                      "value": 2,
                    },
                  ]
                }
              >
                <div
                  data-mock="SubMenu"
                />
              </SubMenu>
            </div>
          </StyledComponent>
        </styled.div>
      </button>
    </StyledComponent>
  </Styled(styled.button)>
</TopMenuItem>
`;

exports[`TopMenuItem simple render 1`] = `
<button
  className="sc-bdVaJa sc-ifAKCX jzIrgm"
  disabled={false}
  onClick={null}
  type="button"
>
  <span
    className="sc-bwzfXH doRSpX"
  />
</button>
`;
